### PR TITLE
Endpoint support auth with permissions

### LINKED ISSUE #45: Gem doesn't support authorization
Gem adapter should support authorization mechanisms as other repositories. Main slice entry point has to accept `Permissions` and `Authentication` objects and use them to construct `SliceAuth` wrapper (see `MavenSlice` or `FilesSlice` for example).
If gem uses HTTP basic authentication, then standard implementation `BasicIdentities` from `artipie/http` module can be used. Otherwise, we have to implement custom authentication implementation in this module.

### PR DESCRIPTION
**Pull request by Sammers21**
Closes #45

### CODE REVIEW

**Comment by Sammers21** [comment]
@olegmoz, review the PR please.

**Review by olegmoz — CHANGES_REQUESTED** [review]
@Sammers21 thanks, please see the comments. It's mostly about lack of tests coverage for the PR. Also there is no integration test to ensure that auth actually works.

**Comment by Sammers21** [comment]
@olegmoz, review the PR once again, please.

File: src/main/java/com/artipie/gem/GemApiKeyIdentity.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public Optional<String> user(final String line,
+        final Iterable<Map.Entry<String, String>> headers) {
+        return new RqHeaders(headers, Authorization.NAME).stream()
+            .findFirst()
+            .map(Base64Decoded::new)
+            .map(dec -> dec.toString().split(":"))
+            .flatMap(cred -> this.auth.user(cred[0].trim(), cred[1].trim()));
+    }
```
[the remark this answers is not in the document]
Comment (Sammers21) [reply src/main/java/com/artipie/gem/GemApiKeyIdentity.java:61]: @olegmoz, should not.

File: src/main/java/com/artipie/gem/GemApiKeyIdentity.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import com.artipie.http.rq.RqHeaders;
+import java.util.Map;
+import java.util.Optional;
+import org.cactoos.text.Base64Decoded;
+
+/**
+ * {@link Identities} implementation fo gem api key decoding.
+ * @since 0.4
+ */
+public final class GemApiKeyIdentity implements Identities {
```
[the remark this answers is not in the document]
Comment (Sammers21) [reply src/main/java/com/artipie/gem/GemApiKeyIdentity.java:38]: @olegmoz, covered with existed tests.

File: src/main/java/com/artipie/gem/GemApiKeyIdentity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sammers21) [reply src/main/java/com/artipie/gem/GemApiKeyIdentity.java]: Proof: 
image

**Review by olegmoz — APPROVED** [review]
@Sammers21 thanks, all is good

**Comment by Sammers21** [comment]
@rultor, release tag=`0.4`

**Comment by 0crat** [comment]
Job `gh:artipie/gem-adapter#59` is not assigned, can't get performer

**Pull request merged**